decode.go

```diff
@@ -570,20 +590,36 @@ func (u *unmarshaler) unmarshalMessage(node *yaml.Node, message proto.Message, f
 		return
 	}
 	// Decode the fields
-	fields := message.ProtoReflect().Descriptor().Fields()
+	msgDesc := message.ProtoReflect().Descriptor()
 	for i := 0; i < len(node.Content); i += 2 {
 		keyNode := node.Content[i]
-		if u.checkKind(keyNode, yaml.ScalarNode) {
-			if forAny && keyNode.Value == atTypeFieldName {
-				continue // Skip the @type field for Any messages
-			}
-			// Get the field Name, JSONName, or Number
-			if field := findField(keyNode.Value, fields); field != nil {
-				valueNode := node.Content[i+1]
-				u.unmarshalField(valueNode, field, message)
-			} else {
-				u.addErrorf(keyNode, "unknown field %#v, expended one of %v", keyNode.Value, getFieldNames(fields))
+		var key string
+		switch keyNode.Kind {
+		case yaml.ScalarNode:
+			key = keyNode.Value
+		case yaml.SequenceNode:
+			if len(keyNode.Content) == 1 && keyNode.Content[0].Kind == yaml.ScalarNode {
+				key = "[" + keyNode.Content[0].Value + "]"
+				break
 			}
+			fallthrough
+		default:
+			u.checkKind(keyNode, yaml.ScalarNode)
+			continue
```

Comment:
The result of `u.checkKind` is never checked. Maybe this was supposed to be this?
```suggestion
			if !u.checkKind(keyNode, yaml.ScalarNode) {
				continue
			}
```

Reply:
this will always fail, so technically doesn't need to be checked

Reply:
Maybe a comment then? Because it's a bit inscrutable/confusing to read as is.

Reply:
done